### PR TITLE
[SPARK-46040][SQL][Python] Update UDTF API for 'analyze' partitioning/ordering columns to support general expressions

### PR DESCRIPTION
**Pull request by dtenedor**
### What changes were proposed in this pull request?

This PR updates the Python user-defined table function (UDTF) API for the `analyze` method to support general expressions for the `partitionBy` and `orderBy` fields of the `AnalyzeResult` class.

For example, the following UDTF specifies to partition by `partition_col / 10` so that all rows with values of this column between 0-9 arrive in the same partition, then all rows with values between 10-19 in the next partition, and so on.

```
@udtf
class TestUDTF:
    def __init__(self):
        self._partition_col = None
        self._count = 0
        self._sum = 0
        self._last = None

    @staticmethod
    def analyze(*args, **kwargs):
        return AnalyzeResult(
            schema=StructType()
            .add("partition_col", IntegerType())
            .add("count", IntegerType())
            .add("total", IntegerType())
            .add("last", IntegerType()),
            partitionBy=[PartitioningColumn("partition_col / 10")],
            orderBy=[
                OrderingColumn(name="input", ascending=True, overrideNullsFirst=False)
            ],
        )

    def eval(self, row: Row):
        self._partition_col = row["partition_col"]
        self._count += 1
        self._last = row["input"]
        if row["input"] is not None:
            self._sum += row["input"]

    def terminate(self):
        yield self._partition_col, self._count, self._sum, self._last
```

### Why are the changes needed?

This lets the UDTF partition by simple references to the columns of the input table just like before, but also gives the option to partition by general expressions based on those columns (just like the explicit `PARTITION BY` and `ORDER BY` clauses in the UDTF call in SQL).

### Does this PR introduce _any_ user-facing change?

Yes, see above.

### How was this patch tested?

This PR includes test coverage.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by dtenedor** [comment]
cc @ueshin @allisonwang-db

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/UserDefinedPythonFunction.scala [hunk 1]
```diff
@@ -190,6 +190,8 @@ class UserDefinedPythonTableFunctionAnalyzeRunner(
 
   override val workerModule = "pyspark.sql.worker.analyze_udtf"
 
+  private lazy val parser = new CatalystSqlParser()
```
Comment (ueshin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/UserDefinedPythonFunction.scala:193]: We should use a parser in `SparkSession`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/UserDefinedPythonFunction.scala [hunk 2]
```diff
@@ -237,33 +239,32 @@ class UserDefinedPythonTableFunctionAnalyzeRunner(
...
-        case 0 =>
-          orderBy.append(SortOrder(UnresolvedAttribute(columnName), direction))
-        case 1 => orderBy.append(
-          SortOrder(UnresolvedAttribute(columnName), direction, NullsFirst, Seq.empty))
-        case 2 => orderBy.append(
-          SortOrder(UnresolvedAttribute(columnName), direction, NullsLast, Seq.empty))
+        case 0 => orderBy.append(SortOrder(parsed, direction))
+        case 1 => orderBy.append(SortOrder(parsed, direction, NullsFirst, Seq.empty))
+        case 2 => orderBy.append(SortOrder(parsed, direction, NullsLast, Seq.empty))
       }
```
Comment (ueshin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/UserDefinedPythonFunction.scala:262]: What happens if the expression contains something like `order_column DESC`?

Comment (dtenedor) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/python/UserDefinedPythonFunction.scala:263]: Anything that fails to parse should fail the query. I added a test case to cover this.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/UserDefinedPythonFunction.scala (hunk 1 above)
Comment (dtenedor) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/python/UserDefinedPythonFunction.scala]: This is done.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/UserDefinedPythonFunction.scala (hunk 2 above)
Comment (ueshin) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/python/UserDefinedPythonFunction.scala:263]: Oh, does it fail? I thought it's a valid expression. NVM, then.

File: sql/core/src/main/scala/org/apache/spark/sql/UDTFRegistration.scala [hunk 3]
```diff
@@ -44,6 +45,7 @@ class UDTFRegistration private[sql] (tableFunctionRegistry: TableFunctionRegistr
          | udfDeterministic: ${udtf.udfDeterministic}
       """.stripMargin)
 
-    tableFunctionRegistry.createOrReplaceTempFunction(name, udtf.builder, "python_udtf")
+    tableFunctionRegistry.createOrReplaceTempFunction(
+      name, udtf.builder(_, new CatalystSqlParser()), "python_udtf")
```
[the remark this answers is not in the document]
Comment (dtenedor) [reply sql/core/src/main/scala/org/apache/spark/sql/UDTFRegistration.scala:49]: Done.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/UserDefinedPythonFunction.scala [hunk 4]
```diff
@@ -106,7 +106,7 @@ case class UserDefinedPythonTableFunction(
     this(name, func, None, pythonEvalType, udfDeterministic)
   }
 
-  def builder(exprs: Seq[Expression]): LogicalPlan = {
+  def builder(exprs: Seq[Expression], parser: ParserInterface): LogicalPlan = {
```
Comment (ueshin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/UserDefinedPythonFunction.scala:109]: This should be `parser: => ParserInterface` to be a by-name argument. https://github.com/apache/spark/pull/43946#discussion_r1408346486

Comment (dtenedor) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/python/UserDefinedPythonFunction.scala]: This is done.

File: sql/core/src/test/scala/org/apache/spark/sql/IntegratedUDFTestUtils.scala [hunk 5]
```diff
@@ -621,10 +623,30 @@ object IntegratedUDFTestUtils extends SQLHelper {
...
+
+  object UDTFInvalidPartitionByOrderByParseError
+    extends TestPythonUDTFPartitionByOrderByBase(
+      partitionBy = "unparsable",
+      orderBy = "input")
+
+  object UDTFInvalidOrderByAscKeyword
+    extends TestPythonUDTFPartitionByOrderByBase(
+      partitionBy = "partition_col",
+      orderBy = "partition_col ASC")
```
Comment (ueshin) [inline sql/core/src/test/scala/org/apache/spark/sql/IntegratedUDFTestUtils.scala:644]: Where are these new tests called?

Comment (dtenedor) [reply sql/core/src/test/scala/org/apache/spark/sql/IntegratedUDFTestUtils.scala:644]: Apologies about that, it seems it was a bad merge from the last PR. I fixed this now, and the tests are used.

File: sql/core/src/test/resources/sql-tests/results/udtf/udtf.sql.out [hunk 6]
```diff
@@ -335,6 +335,37 @@ org.apache.spark.sql.AnalysisException
...
+SELECT * FROM InvalidTestPythonUDTFOrderByAscKeyword(TABLE(t2))
+-- !query schema
+struct<>
+-- !query output
+org.apache.spark.sql.AnalysisException
+{
+  "errorClass" : "UNRESOLVABLE_TABLE_VALUED_FUNCTION",
+  "sqlState" : "42883",
+  "messageParameters" : {
+    "name" : "`InvalidTestPythonUDTFOrderByAscKeyword`"
```
Comment (ueshin) [inline sql/core/src/test/resources/sql-tests/results/udtf/udtf.sql.out:357]: Is this expected? From the class name, it seems not expected?

Comment (dtenedor) [reply sql/core/src/test/resources/sql-tests/results/udtf/udtf.sql.out]: Apologies, I forgot to update the UDTF name. Fixed this now, the error message works as expected.

Comment (ueshin) [reply sql/core/src/test/resources/sql-tests/results/udtf/udtf.sql.out]: Ah, then, the updated test seems to be duplicated now:

```sql
SELECT * FROM UDTFInvalidOrderByAscKeyword(TABLE(t2));
```

and the latter one failed with `ParseException` by missing `;` at the end.

Comment (dtenedor) [reply sql/core/src/test/resources/sql-tests/results/udtf/udtf.sql.out]: Thanks for pointing that out, fixed it again!

**Review by ueshin — APPROVED** [review]
LGTM, pending tests.

**Comment by ueshin** [comment]
Thanks! merging to master.

**Pull request closed without merging**